# Assignment 1: Neural Networks
# BA 64061-003
# Madeline Witzeman

# First Step: Loading the IMDb dataset

In [1]:
import tensorflow
from tensorflow.keras.datasets import imdb
(train_data, train_labels), (test_data, test_labels) = imdb.load_data(
    num_words=10000)

# Second Step: Recreating the IMDb example studied in class so I have a baseline model to compare against when trying to improve model performance

## Preparing the data

Encoding the integer sequences via multi-hot encoding so the data is ready to be fed into a neural network

In [2]:
import numpy as np
def vectorize_sequences(sequences, dimension=10000):
    results = np.zeros((len(sequences), dimension))
    for i, sequence in enumerate(sequences):
        for j in sequence:
            results[i, j] = 1.
    return results
x_train = vectorize_sequences(train_data)
x_test = vectorize_sequences(test_data)

Vectorizing labels:

In [3]:
y_train = np.asarray(train_labels).astype("float32")
y_test = np.asarray(test_labels).astype("float32")

## Building the model

Defining the model:

In [4]:
from tensorflow import keras
from tensorflow.keras import layers

model = keras.Sequential([
    layers.Dense(16, activation="relu"),
    layers.Dense(16, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])

Compiling the model:

In [5]:
model.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])

## Validating approach

Setting aside validation set:

In [6]:
x_val = x_train[:10000]
partial_x_train = x_train[10000:]
y_val = y_train[:10000]
partial_y_train = y_train[10000:]

Training the model (using epoch = 4, determined as best epoch value in the textbook):

In [7]:
history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=4,
                    batch_size=512,
                    validation_data=(x_val, y_val))

Epoch 1/4
30/30 [==============================] - 4s 87ms/step - loss: 0.5538 - accuracy: 0.7741 - val_loss: 0.4346 - val_accuracy: 0.8538
Epoch 2/4
30/30 [==============================] - 1s 37ms/step - loss: 0.3514 - accuracy: 0.8915 - val_loss: 0.3323 - val_accuracy: 0.8768
Epoch 3/4
30/30 [==============================] - 1s 38ms/step - loss: 0.2628 - accuracy: 0.9113 - val_loss: 0.2960 - val_accuracy: 0.8835
Epoch 4/4
30/30 [==============================] - 1s 37ms/step - loss: 0.2099 - accuracy: 0.9296 - val_loss: 0.3091 - val_accuracy: 0.8732


In [10]:
results = model.evaluate(x_test, y_test)
results

782/782 [==============================] - 2s 2ms/step - loss: 0.3193 - accuracy: 0.8756


[0.3192662000656128, 0.8755599856376648]

Using two hidden layers, 16 hidden units, binary_crossentropy loss function, relu activation, and epoch = 4 yields 88.34% accuracy on the validation dataset and 87.69% accuracy on the test dataset, on average.

Since the neural network I'm utilizing implements stochastic learning, the model produces different validation and test accuracies each time it is run. To arrive at the validation and test accuracies for each model, I will take the average of 5 runs:

1.   Val: .8837; Test: .8789
2.   Val: .8762; Test: .8666
3.   Val: .8846; Test: .8785
4.   Val: .8879; Test: .8847
5.   Val: .8847; Test: .8756

Average - Val: .8834; Test: .8769



Now that I know the validation and test accuracy of the baseline model, I'm going to try various approaches to improve the model's performance by altering one component at a time.

# Next approach: Using one hidden layer instead of two hidden layers

## Retraining the model

In [9]:
model_2 = keras.Sequential([
    layers.Dense(16, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])
model_2.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])
model_2.fit(partial_x_train, partial_y_train, epochs=4, batch_size=512, validation_data=(x_val, y_val))
results = model_2.evaluate(x_test, y_test)
results

Epoch 1/4
30/30 [==============================] - 4s 98ms/step - loss: 0.5246 - accuracy: 0.7867 - val_loss: 0.4172 - val_accuracy: 0.8623
Epoch 2/4
30/30 [==============================] - 1s 45ms/step - loss: 0.3566 - accuracy: 0.8870 - val_loss: 0.3439 - val_accuracy: 0.8783
Epoch 3/4
30/30 [==============================] - 1s 37ms/step - loss: 0.2827 - accuracy: 0.9114 - val_loss: 0.3057 - val_accuracy: 0.8857
Epoch 4/4
782/782 [==============================] - 2s 3ms/step - loss: 0.3101 - accuracy: 0.8757


[0.31009921431541443, 0.8757200241088867]

## Results

Using only one hidden layer increases the validation accuracy to 88.88% and the test accuracy to 88.26%.

Run data:


1.   Val: .8868; Test: .8786
2.   Val: .8894; Test: .8839
3.   Val: .8894; Test: .8836
4.   Val: .8890; Test: .8839
5.   Val: .8894; Test: .8832

Average - Val: .8888; Test: .8826



# Using three hidden layers instead of two hidden layers

## Retraining the model

In [13]:
model_3 = keras.Sequential([
    layers.Dense(16, activation="relu"),
    layers.Dense(16, activation="relu"),
    layers.Dense(16, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])
model_3.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])
model_3.fit(partial_x_train, partial_y_train, epochs=4, batch_size=512, validation_data=(x_val, y_val))
results = model_3.evaluate(x_test, y_test)
results

Epoch 1/4
30/30 [==============================] - 3s 69ms/step - loss: 0.5717 - accuracy: 0.7588 - val_loss: 0.4415 - val_accuracy: 0.8590
Epoch 2/4
30/30 [==============================] - 1s 35ms/step - loss: 0.3552 - accuracy: 0.8870 - val_loss: 0.3224 - val_accuracy: 0.8837
Epoch 3/4
30/30 [==============================] - 1s 36ms/step - loss: 0.2570 - accuracy: 0.9134 - val_loss: 0.3015 - val_accuracy: 0.8820
Epoch 4/4
782/782 [==============================] - 2s 2ms/step - loss: 0.2959 - accuracy: 0.8813


[0.29587337374687195, 0.8813199996948242]

## Results

Using three hidden layers slightly increased the validation accuracy to 88.52% and the test accuracy to 87.94%.

Run data:


1.   Val: .8890; Test: .8827
2.   Val: .8900; Test: .8840
3.   Val: .8879; Test: .8822
4.   Val: .8735; Test: .8668
5.   Val: .8857; Test: .8813

Average - Val: .8852; Test: .8794



## Using one hidden layer produced the best validation and test performance

# Increasing hidden units to 32 instead of 16

## Retraining the model

In [12]:
model_4 = keras.Sequential([
    layers.Dense(32, activation="relu"),
    layers.Dense(32, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])
model_4.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])
model_4.fit(partial_x_train, partial_y_train, epochs=4, batch_size=512, validation_data=(x_val, y_val))
results = model_4.evaluate(x_test, y_test)
results

Epoch 1/4
30/30 [==============================] - 3s 87ms/step - loss: 0.5152 - accuracy: 0.7737 - val_loss: 0.3769 - val_accuracy: 0.8661
Epoch 2/4
30/30 [==============================] - 2s 53ms/step - loss: 0.3044 - accuracy: 0.8942 - val_loss: 0.3199 - val_accuracy: 0.8688
Epoch 3/4
30/30 [==============================] - 2s 53ms/step - loss: 0.2252 - accuracy: 0.9209 - val_loss: 0.3100 - val_accuracy: 0.8723
Epoch 4/4
782/782 [==============================] - 2s 3ms/step - loss: 0.2904 - accuracy: 0.8825


[0.2903912663459778, 0.8824800252914429]

## Results

Using 32 hidden units in hidden layers slightly decreased the validation accuracy to 88.09% and the test accuracy to 87.31%. Since performance decreased for both the validation and test sets, I won't use 32 hidden units moving forward.




Run data:


1.   Val: .8815; Test: .8712
2.   Val: .8880; Test: .8822
3.   Val: .8607; Test: .8486
4.   Val: .8869; Test: .8810
5.   Val: .8874; Test: .8825

Average - Val: .8809; Test: .8731



# Increasing hidden units to 64 instead of 16

## Retraining the model

In [10]:
model_5 = keras.Sequential([
    layers.Dense(64, activation="relu"),
    layers.Dense(64, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])
model_5.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])
model_5.fit(partial_x_train, partial_y_train, epochs=4, batch_size=512, validation_data=(x_val, y_val))
results = model_5.evaluate(x_test, y_test)
results

Epoch 1/4
30/30 [==============================] - 5s 123ms/step - loss: 0.5112 - accuracy: 0.7633 - val_loss: 0.3494 - val_accuracy: 0.8680
Epoch 2/4
30/30 [==============================] - 2s 66ms/step - loss: 0.2991 - accuracy: 0.8824 - val_loss: 0.2835 - val_accuracy: 0.8886
Epoch 3/4
30/30 [==============================] - 2s 67ms/step - loss: 0.2219 - accuracy: 0.9162 - val_loss: 0.2799 - val_accuracy: 0.8854
Epoch 4/4
782/782 [==============================] - 4s 5ms/step - loss: 0.3251 - accuracy: 0.8686


[0.32506248354911804, 0.8685600161552429]

## Results

Using 64 hidden units in hidden layers decreased the validation accuracy notably to 87.13% and the test accuracy to 86.06%. It appears that adding hidden units decreases model performance when using 4 epoches. Given this information, I won't use 64 hidden units moving forward.

Run data:


1.   Val: .8646; Test: .8535
2.   Val: .8882; Test: .8830
3.   Val: .8613; Test: .8478
4.   Val: .8637; Test: .8501
5.   Val: .8785; Test: .8686

Average - Val: .8713; Test: .8606



## Using 16 hidden units in hidden layers produced the best validation and test performance

# Using MSE loss function instead of binary_crossentropy

## Retraining the model

In [11]:
model_6 = keras.Sequential([
    layers.Dense(16, activation="relu"),
    layers.Dense(16, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])
model_6.compile(optimizer="rmsprop",
              loss="mse",
              metrics=["accuracy"])
model_6.fit(partial_x_train, partial_y_train, epochs=4, batch_size=512, validation_data=(x_val, y_val))
results = model_6.evaluate(x_test, y_test)
results

Epoch 1/4
30/30 [==============================] - 4s 108ms/step - loss: 0.1732 - accuracy: 0.7670 - val_loss: 0.1200 - val_accuracy: 0.8658
Epoch 2/4
30/30 [==============================] - 2s 53ms/step - loss: 0.0980 - accuracy: 0.8912 - val_loss: 0.0993 - val_accuracy: 0.8755
Epoch 3/4
30/30 [==============================] - 1s 35ms/step - loss: 0.0747 - accuracy: 0.9136 - val_loss: 0.0869 - val_accuracy: 0.8884
Epoch 4/4
782/782 [==============================] - 2s 2ms/step - loss: 0.0881 - accuracy: 0.8828


[0.08807112276554108, 0.8828399777412415]

## Results

Using the mean squared error (mse) loss function kept the validation accuracy around 88.33% and slightly increased the test accuracy to 87.88%. Given this information, I may consider using the mse loss function moving forward.

Run data:


1.   Val: .8727; Test: .8726
2.   Val: .8853; Test: .8809
3.   Val: .8840; Test: .8766
4.   Val: .8861; Test: .8812
5.   Val: .8883; Test: .8828

Average - Val: .8833; Test: .8788



# Using tanh activation instead of relu

## Retraining the model

In [9]:
model_7 = keras.Sequential([
    layers.Dense(16, activation="tanh"),
    layers.Dense(16, activation="tanh"),
    layers.Dense(1, activation="sigmoid")
])
model_7.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])
model_7.fit(partial_x_train, partial_y_train, epochs=4, batch_size=512, validation_data=(x_val, y_val))
results = model_7.evaluate(x_test, y_test)
results

Epoch 1/4
30/30 [==============================] - 3s 77ms/step - loss: 0.5149 - accuracy: 0.7959 - val_loss: 0.4037 - val_accuracy: 0.8525
Epoch 2/4
30/30 [==============================] - 1s 36ms/step - loss: 0.3155 - accuracy: 0.8953 - val_loss: 0.3099 - val_accuracy: 0.8787
Epoch 3/4
30/30 [==============================] - 1s 33ms/step - loss: 0.2312 - accuracy: 0.9225 - val_loss: 0.2852 - val_accuracy: 0.8848
Epoch 4/4
782/782 [==============================] - 2s 3ms/step - loss: 0.3033 - accuracy: 0.8741


[0.30325186252593994, 0.8740800023078918]

## Results

Using the tanh activation slightly increased the validation accuracy to 88.48% and test accuracy to 87.78%. Given this information, I may consider using tanh activation moving forward.

Run data:


1.   Val: .8841; Test: .8768
2.   Val: .8850; Test: .8777
3.   Val: .8890; Test: .8822
4.   Val: .8846; Test: .8784
5.   Val: .8812; Test: .8741

Average - Val: .8848; Test: .8778



# Attempt at optimizing the model based on prior results

# Using one hidden layer + 16 hidden units + mse loss function + tanh activation to explore model performance

## Retraining the model

In [9]:
model_8 = keras.Sequential([
    layers.Dense(16, activation="tanh"),
    layers.Dense(1, activation="sigmoid")
])
model_8.compile(optimizer="rmsprop",
              loss="mse",
              metrics=["accuracy"])
model_8.fit(partial_x_train, partial_y_train, epochs=4, batch_size=512, validation_data=(x_val, y_val))
results = model_8.evaluate(x_test, y_test)
results

Epoch 1/4
30/30 [==============================] - 3s 69ms/step - loss: 0.1691 - accuracy: 0.7848 - val_loss: 0.1332 - val_accuracy: 0.8455
Epoch 2/4
30/30 [==============================] - 1s 36ms/step - loss: 0.1049 - accuracy: 0.8916 - val_loss: 0.1021 - val_accuracy: 0.8814
Epoch 3/4
30/30 [==============================] - 1s 36ms/step - loss: 0.0823 - accuracy: 0.9118 - val_loss: 0.0989 - val_accuracy: 0.8733
Epoch 4/4
782/782 [==============================] - 2s 3ms/step - loss: 0.0912 - accuracy: 0.8844


[0.09115107357501984, 0.884440004825592]

## Results

This model consistently had the highest performance with a 88.67% validation accuracy and 88.31% test accuracy. It appears that adjusting the base model using the components that improved performance earlier in the analysis collectively made the model better.

Run data:

1.   Val: .8878; Test: .8852
2.   Val: .8871; Test: .8827
3.   Val: .8893; Test: .8851
4.   Val: .8880; Test: .8830
5.   Val: .8812; Test: .8796

Average - Val: .8867; Test: .8831

